### ISSUE TITLE
Consider writing out-of-class operators

### ISSUE THREAD

**Issue by eccentricOrange** [eccentricOrange/c-cpp-mirror#7]
Instead of writing the operators as class member functions, you can also make them out-of-class [`friend` functions](https://www.learncpp.com/cpp-tutorial/overloading-the-arithmetic-operators-using-friend-functions/):

```cpp
template<…>
class Matrix {
    …
    friend Matrix operator+(const Matrix& lhs, const Matrix& rhs);
    …
};

template<std::size_t R, std::size_t C>
Matrix<R, C> operator+(const Matrix<R, C>& lhs, const Matrix<R, C>& rhs) {
    Matrix<R, C> result;
    …
        result[row][column] = lhs[row][column] + rhs[row][column];
    …
    return result;
}
```

You might not see the point of this, but it has several advantages. In particular, consider that you might want to allow multiplying a matrix by a scalar value. You could add an overload of the member function `operator*()`:

```cpp
Matrix operator*(double value) {…}
```

That would allow you to write:

```cpp
Matrix<…> matrix(…);
auto matrix2 = matrix * 2;
```

But it would not allow you to write:

```cpp
auto matrix2 = 2 * matrix;
```

And there is no way to achieve that with a member function. However, the `friend` operators get two parameters, so then you can do:

```cpp
friend Matrix operator*(const Matrix& lhs, double rhs);
friend Matrix operator*(double lhs, const Matrix& rhs);
```

In this case you can easily define one overload in terms of the other in order to avoid code duplication:

```cpp
template<std::size_t R, std::size_t C>
friend Matrix<R, C> operator*(double lhs, const Matrix<R, C>& rhs) {
    return rhs * lhs;
}
```

Apart from the binary operators themselves, you can also make members like `getDeterminant()` friends instead:

```cpp
friend double getDeterminant(const Matrix& matrix);
```

You would then need to call it like `getDeterminant(matrix)` instead of `matrix.getDeterminant()`, but that's similar to how you use `std::abs()` and other standard math functions.

**Comment by eccentricOrange** [comment]
Fixed in #16

**Comment by eccentricOrange** [comment]
Fixed in #16
